doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,537 @@
+
+We currently surface details about both Sources and Tables through various system catalog
+tables in both the `mz_catalog` and `mz_internal` schemas. These expose information
+about the objects themselves as well as statistics and statuses emitted from the active dataflows for each source.
+
+We would make the following updates to provide users and Materialize employees the same
+ability to introspect source-fed tables that they currently have for subsources.
+
+#### [mz_catalog.mz_tables](https://materialize.com/docs/sql/system-catalog/mz_catalog/#mz_tables)
+
+`CREATE TABLE .. FROM SOURCE` statements would be added to this catalog table.
+
+Two new columns would be added:
+
+`data_source_type`: distinguishes these tables
+from 'normal' tables that are written to directly by users. The value of this will
+be one of `'direct'` or `'source'`.
+
+`data_source_id`: a nullable column that contains the `id` of the data source,
+if there is one. In the case of `CREATE TABLE .. FROM SOURCE` statements this would
+contain the `id` of the primary source.
+
+#### [mz_catalog.mz_sources](https://materialize.com/docs/sql/system-catalog/mz_catalog/#mz_sources)
+
+`CREATE TABLE .. FROM SOURCE` statements will be added to this catalog table, despite
+them being 'tables', since this catalog table contains columns that are unique to
+source-related objects (e.g. `connection_id`).
+
+We will add the variant `'table'` to the `type` column which is currently a `text` column
+containing one of:
+`'kafka', 'mysql', 'postgres', 'load-generator', 'progress', or 'subsource'.`
+
+Once all customer's sources have been migrated to the new model, the `'subsource'` variant will be removed from the `type` column.
+
+Open Question: Will it be confusing to users that these source-fed tables will appear
+in both the `mz_sources` and `mz_tables` catalog tables?
+
+#### [mz_internal.mz_postgres_source_tables](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_postgres_source_tables) and [mz_internal.mz_mysql_source_tables](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_mysql_source_tables)
+
+Each row in these tables contains a mapping from a `subsource` `id` to the external
+`schema` and `table` of the Postgres/MySQL table being ingested into that subsource.
+
+These will be updated to include the same mappings for `CREATE TABLE .. FROM SOURCE`
+table objects to the external `schema` and `table` they are linked to.
+
+#### [mz_internal.mz_source_statistics](https://materialize.com/docs/sql/system-catalog/mz_internal/#mz_source_statistics)
+
+This table does not require any changes to its schema. However it is important to highlight
+that since multiple tables will be able to ingest from the same upstream external reference,
+the values reported in the following fields will be representative of the total number of
+records output to all tables rather than the number of rows ingested from the external references
+of the upstream system:
```

Comment:
I didn't quite follow this! Will there continue to be a row in this table for each subsource _and_ the top-level source, or just the top-level source?

Reply:
Same confusion as Nikhil.

Reply:
Ahh I thought this table only had rows for top-level sources, my mistake.

So it seems that nothing needs to change in this table, we will just have table-fed sources listed here just like subsourcs are since the metrics for each are essentially the same

Reply:
Yeah, I think that's right. Separately, it is massively confusing what gets rolled up into the parent source and what doesn't—some stats are duplicated in both places; some stats only appear in one both not the other—but that's probably not something to bite of fixing here. I think the status quo is just fine.